File: src/main/java/com/itsci/mju/maebanjumpen/model/PartyRole.java
// File: src/main/java/com/itsci/mju/maebanjumpen/model/PartyRole.java
package com.itsci.mju.maebanjumpen.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import jakarta.persistence.*;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.EqualsAndHashCode;

@Entity
@DiscriminatorColumn(name = "DTYPE", discriminatorType = DiscriminatorType.STRING)
@NoArgsConstructor
//@Inheritance(strategy = InheritanceType.JOINED)
@Data
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
@JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type"
)
@JsonSubTypes({
        @JsonSubTypes.Type(value = Member.class, name = "member"),
        @JsonSubTypes.Type(value = Hirer.class, name = "hirer"),
        @JsonSubTypes.Type(value = Housekeeper.class, name = "housekeeper"),
        @JsonSubTypes.Type(value = Admin.class, name = "admin"),
        @JsonSubTypes.Type(value = AccountManager.class, name = "accountManager"),
})
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public abstract class PartyRole {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    private Integer id;

    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "person_id", referencedColumnName = "person_id", nullable = false, unique = true)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private Person person;

}

File: src/main/kotlin/com/itsci/mju/maebanjumpen/entity/PartyRole.kt
package com.itsci.mju.maebanjumpen.entity

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import jakarta.persistence.*

@Entity
@DiscriminatorColumn(name = "DTYPE", discriminatorType = DiscriminatorType.STRING)
@JsonIgnoreProperties("hibernateLazyInitializer", "handler")
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type"
)
@JsonSubTypes(
    JsonSubTypes.Type(value = Member::class, name = "member"),
    JsonSubTypes.Type(value = Hirer::class, name = "hirer"),
    JsonSubTypes.Type(value = Housekeeper::class, name = "housekeeper"),
    JsonSubTypes.Type(value = Admin::class, name = "admin"),
    JsonSubTypes.Type(value = AccountManager::class, name = "accountManager")
)
abstract class PartyRole(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = 0,

    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "person_id", referencedColumnName = "personId", nullable = false, unique = true)
    open var person: Person? = null
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PartyRole) return false
        return id != null && id == other.id
    }

    override fun hashCode(): Int = id?.hashCode() ?: 0
}

